### ISSUE TITLE
torch.nn.CTCLoss not reproducible on GPU

### ISSUE THREAD

**Issue by stefbraun** [pytorch/pytorch#17798]
Updated again.

The inclusion of the CTC loss function in pytorch is great. However, the loss may be deterministic or non-deterministic under some conditions. In typical speech training scenarios, it could actually happen that for some batches the CDUNN algorithm is used (which can be set to deterministic or non-deterministic modes), and for some batches the CUDA algorithm is used (which is always non-deterministic).

| Device | `USE_CUDNN` | `torch.backends.cudnn.deterministic` | Code | Deterministic |
|---|---|---|---|---|
| CPU | - | -| [ATen/native/LossCTC.cpp](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/LossCTC.cpp)  | Yes | 
| GPU | False | - | [ATen/native/cuda/LossCTC.cu](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cuda/LossCTC.cu) | No |
| GPU | True | True  |  [ATen/native/cudnn/LossCTC.cpp](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cudnn/LossCTC.cpp) | Yes |
| GPU | True | False | [ATen/native/cudnn/LossCTC.cpp](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cudnn/LossCTC.cpp) | No|

- `USE_CUDNN` is a boolean variable re-generated for every batch [here](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/LossCTC.cpp#L336-L355)
EDIT: false alarm, sorry there was a bug in my code, please delete

I have attached a code sample for reproducing the non-deterministic loss function. Even though the difference here is small, it gets larger in training of speech recognition models. Anyways, for research purposes, exact reproducibility is rather important. 

Would it be possible to introduce a deterministic mode in the CUDA version? To the best of my knowledge, this need an alternative for `atomicAdd` in [`LossCTC.cu`](https://github.com/pytorch/pytorch/blob/29f096cc70cf9cc1a317ae7107228215b7dde60b/aten/src/ATen/native/cuda/LossCTC.cu#L409-L412).

Output:
```
7402
Trial 0 ::: Device cuda ::: Iter 99 ::: net_csum -1.2945041656 ::: log_prob_checksum -500026.3125000000 ::: loss 10.0990858078 ::: inc_loss 1045.5471239090
Trial 1 ::: Device cuda ::: Iter 99 ::: net_csum -1.2944996357 ::: log_prob_checksum -500026.2500000000 ::: loss 10.0990848541 ::: inc_loss 1045.5470838547
Trial 2 ::: Device cuda ::: Iter 99 ::: net_csum -1.2945439816 ::: log_prob_checksum -500026.2500000000 ::: loss 10.0990819931 ::: inc_loss 1045.5470523834
Trial 0 ::: Device cpu  ::: Iter 99 ::: net_csum 2.3939034939 ::: log_prob_checksum -499604.8750000000 ::: loss 10.1452789307 ::: inc_loss 1050.8508319855
Trial 1 ::: Device cpu  ::: Iter 99 ::: net_csum 2.3939034939 ::: log_prob_checksum -499604.8750000000 ::: loss 10.1452789307 ::: inc_loss 1050.8508319855
Trial 2 ::: Device cpu  ::: Iter 99 ::: net_csum 2.3939034939 ::: log_prob_checksum -499604.8750000000 ::: loss 10.1452789307 ::: inc_loss 1050.8508319855
```
Code:
```
import torch
import torch.nn as nn

def train_with_ctc(trial, device):
    
    # Random seed
    torch.manual_seed(0)

    # Data
    torch.manual_seed(0)
    inp=torch.randn(500,16,10, device=device)    
    targets = torch.randint(1, 20, (sum(range(100,116)),), dtype=torch.int)
    input_lengths = torch.full((16,), 500, dtype=torch.long)
    target_lengths = torch.arange(100,116, dtype=torch.long)
    
    # Network
    net=nn.Sequential(nn.Linear(10,20), torch.nn.LogSoftmax(2))
    net.to(device)
    optimizer=torch.optim.Adam(net.parameters())
    ctc_loss = nn.CTCLoss(blank=19) # use cuDNN: blank=0, use CUDA: blank !=0 (only true when other parameters are not changed)
        
    # Training loop
    inc_loss=0
    for it in range(100):
        optimizer.zero_grad()
        log_probs=net(inp)
        loss = ctc_loss(log_probs, targets, input_lengths, target_lengths)
        loss.backward()
        optimizer.step()
        inc_loss+=float(loss)

    # Print
    torch.cuda.synchronize()
    net_csum=torch.nn.utils.parameters_to_vector(net.parameters()).sum()
    print('Trial {} ::: Device {:4} ::: Iter {} ::: net_csum {:.10f} ::: log_prob_checksum {:.10f} ::: loss {:.10f} ::: inc_loss {:.10f}'.format(trial, device, it, net_csum,log_probs.sum(), float(loss), inc_loss))

if __name__ == '__main__':

    # Get cuDNN version
    print(torch.backends.cudnn.version())
    
    # Toggle cuDNN backend    
    # torch.backends.cudnn.deterministic = True # change to True and blank label to 0 to toggle CUDNN_CTC_LOSS_ALGO_DETERMINISTIC / CUDNN_CTC_LOSS_ALGO_NON_DETERMINISTIC in pytorch/aten/src/ATen/native/cudnn/LossCTC.cpp
    # torch.backends.cudnn.benchmark = False
    
    # Train with CTC
    for device in ['cuda', 'cpu']:
        for trial in range(3):
            train_with_ctc(trial, device)
        
```

**Comment by stefbraun** [comment]
Correct code:

import torch
import torch.nn as nn

torch.backends.cudnn.deterministic = True
torch.backends.cudnn.benchmark = False

all_devices = ['cpu', 'cuda']

for device in all_devices:
    torch.manual_seed(0)
    log_probs = torch.randn(50, 16, 20, device=device).log_softmax(2).detach().requires_grad_()
    targets = torch.randint(1, 20, (16, 30), dtype=torch.long)
    input_lengths = torch.full((16,), 50, dtype=torch.long)
    target_lengths = torch.randint(10, 30, (16,), dtype=torch.long)

    for iter in range(3):
        ctc_loss = nn.CTCLoss()
        loss = ctc_loss(log_probs, targets, input_lengths, target_lengths)
        loss.backward()

        print('Device {} ::: Iter {} ::: checksum {}'.format(device, iter, log_probs.grad.sum()))
        log_probs.grad.data.zero_()

Output:
Device cpu ::: Iter 0 ::: checksum 5.617737770080566e-06
Device cpu ::: Iter 1 ::: checksum 5.617737770080566e-06
Device cpu ::: Iter 2 ::: checksum 5.617737770080566e-06
Device cuda ::: Iter 0 ::: checksum -6.280839443206787e-06
Device cuda ::: Iter 1 ::: checksum -6.280839443206787e-06
Device cuda ::: Iter 2 ::: checksum -6.280839443206787e-06

**Comment by ezyang** [comment]
cc @t-vi

**Comment by ezyang** [comment]
> Would it be possible to introduce a deterministic mode in the CUDA version?

This seems reasonable.

WarpCTC is deterministic on GPU, so we might be able to reuse the implementation there.

@t-vi, do you have any better ideas for how to deal with this?

**Comment by t-vi** [comment]
It's documented that the loss backward is non-deterministic currently.

So basically, the atomicAdd is used in an `index_add`-like fashion. The obvious replacement would be something like the deterministic `index_add` of embedding backward.
Personally, I don't think using the WarpCTC implementation would be an improvement.

We do have a bug open somewhere regarding determinism.

Obviously we'd be happy if anyone wants to fix this.

**Comment by ezyang** [comment]
I can commit to reviewing a PR that adds a deterministic mode to this kernel. Not sure if we should make it deterministic by default; depends on the performance cost.

**Comment by stefbraun** [comment]
The [note](https://pytorch.org/docs/stable/nn.html#torch.nn.CTCLoss) in the docs concerning the determinism could be improved:

"In some circumstances when using the CUDA backend with CuDNN, this operator may select a nondeterministic algorithm to increase performance. If this is undesirable, you can try to make the operation deterministic (potentially at a performance cost) by setting torch.backends.cudnn.deterministic = True. Please see the notes on Reproducibility for background."

- "Some circumstances" is a bit imprecise
- "CUDA backend with CuDNN" - when using a NVIDIA GPU, there seems to be the CUDA implementation and two CuDNN implementations (`CUDNN_CTC_LOSS_ALGO_DETERMINISTIC` and `CUDNN_CTC_LOSS_ALGO_NON_DETERMINISTIC`) that may be used

**Comment by mruberry** [comment]
Updated issue's title. Note that torch.nn.CTCLoss is correctly documented as nondeterministic here: 

https://pytorch.org/docs/master/generated/torch.use_deterministic_algorithms.html?highlight=use_deterministic_algorithms#torch.use_deterministic_algorithms.

**Comment by HANDSOMEJACKANDY** [comment]
It seems that the torch.nn.CTCLoss with CUDNN backend activated still has non-deterministic backward gradient. Is there any update on this feature request?

**Comment by t-vi** [comment]
I would not hold my breath. The backward non-determinism is from what is essentially a scatter-accumulation operation (similar to scatter_add), so one would need to implement deterministic scanning. This can be done but once you have thought it through, the details are boring and tedious, so you would need to have a lot of determination to fix this. If I were to work on things like that, I'd probably do scatter_add first, too.
